### DISCUSSION TITLE
Integration variables derivatives in generalised behaviours

### DISCUSSION THREAD

**Discussion by mhpierre** [thelfer/tfel#521] [Q&A]
Dear Thomas,

I'm working with thermoporoplasticity in MFront, and I have a small interrogation about the calculation of derivatives of the integration variables with respect to the external variables. From what I understand, there are two possible ways to do it: 

- By calling the blocks of the invert of the Jacobian `iJ_x_y` directly and calculating "by hand",
- By using the `getIntegrationVariablesDerivatives` method.

I have an elastic strain and a plastic multiplier, I define the right derivatives of the implicit equations with respect to the external variables, as so:

```
∂fε_el∕∂Δεᵗᵒ = - I₄ ;
∂fλₚ∕∂Δpₗ  = (β-b)⋅∂f∕∂p / (E⋅E);
∂fλₚ∕∂ΔT    = -κ ⋅∂f∕∂p / (E⋅E);
```

where other derivatives are zero. 
For the first solution, it works as intended:

```
Stensor4 ∂Δε_el∕∂Δεᵗᵒ = -( iJ_ε_el_ε_el ⋅∂fε_el∕∂Δεᵗᵒ ) ;
Stensor ∂Δε_el∕∂Δpₗ  = -( iJ_ε_el_λₚ ⋅∂fλₚ∕∂Δpₗ ) ;
Stensor ∂Δε_el∕∂ΔT  = -( iJ_ε_el_λₚ ⋅∂fλₚ∕∂ΔT ) ;
```

However if I use the `getIntegrationVariablesDerivatives` method, as so:

```
Stensor4 ∂Δε_el∕∂Δεᵗᵒ ;
Stensor  ∂Δε_el∕∂Δpₗ ;
Stensor  ∂Δε_el∕∂ΔT ;
        
getIntegrationVariablesDerivatives_εᵗᵒ(∂Δε_el∕∂Δεᵗᵒ);
getIntegrationVariablesDerivatives_pₗ(∂Δε_el∕∂Δpₗ);
getIntegrationVariablesDerivatives_T(∂Δε_el∕∂ΔT);
```

I get a warning on compilation that variables `∂fλₚ∕∂Δpₗ` and `∂fλₚ∕∂ΔT` are being set but not used. Am I misunderstanding the way `getIntegrationVariablesDerivatives` works?

**Comment by thelfer** [comment]
Those variables are views to blocks of a zero initialized matrix. If those derivatives are null, you don't need to use those variables. In this case, you can explicitly discard them as follows:

```c++
static_cast<void>(∂fλₚ∕∂Δpₗ);
```
